x-pack/plugins/security_solution/public/common/utils/global_query_string/index.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as H from 'history';
+import { parse, ParsedQuery, stringify } from 'query-string';
+import { useCallback, useEffect, useMemo } from 'react';
+
+import { url } from '@kbn/kibana-utils-plugin/public';
+import { isEmpty, pickBy } from 'lodash/fp';
+import { useHistory, useLocation } from 'react-router-dom';
+import { useDispatch } from 'react-redux';
+import {
+  decodeRisonUrlState,
+  encodeRisonUrlState,
+  getParamFromQueryString,
+  getQueryStringFromLocation,
+} from '../../components/url_state/helpers';
+import { useShallowEqualSelector } from '../../hooks/use_selector';
+import { globalUrlParamActions, globalUrlParamSelectors } from '../../store/global_url_param';
+import { useRouteSpy } from '../route/use_route_spy';
+import { getLinkInfo } from '../../links';
+
+/**
+ * Adds urlParamKey and the initial value to redux store.
+ *
+ * Please call this hook at the highest possible level of the rendering tree.
+ * So it is only called when the application starts instead of on every page.
+ *
+ * @param urlParamKey Must not change.
+ * @param onInitialize Called once when initializing.
+ */
+export const useInitializeUrlParam = <State>(
+  urlParamKey: string,
+  /**
+   * @param state Decoded URL param value.
+   */
+  onInitialize: (state: State | null) => void
+) => {
+  const dispatch = useDispatch();
+  const { search } = useLocation();
+
+  useEffect(() => {
+    const initialValue = getParamFromQueryString(getQueryStringFromLocation(search), urlParamKey);
+
+    dispatch(
+      globalUrlParamActions.registerUrlParam({
+        key: urlParamKey,
+        initialValue: initialValue ?? null,
+      })
+    );
+
+    // execute consumer initialization
+    onInitialize(decodeRisonUrlState<State>(initialValue ?? undefined));
+
+    return () => {
+      dispatch(globalUrlParamActions.deregisterUrlParam({ key: urlParamKey }));
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- It must run only once when the application is initializing.
```

Comment:
Sorry just saw this, is there anyway we can rework this not use this. Otherwise, it might actually be better to go functional to enforce it runs once. Per discussion with @janmonschke and @andrew-goldstein  it might be worth looking at an implementation such as: https://beta.reactjs.org/learn/synchronizing-with-effects#not-an-effect-initializing-the-application

Reply:
I am a bit confused by the comment. If you are referring to the new react 18 `useEffect` behavior that calls an effect twice, that would not be a problem. The hook would register, unregister and register again.

Reply:
I think it is no problem if the effect runs multiple times, the state will stay consistent, so we don't really need to enforce this effect to be executed only once. The effect dependencies are empty only for performance purposes, but we could add the missing dependencies and let it register/unregister as many times as it needs.

Nonetheless, I agree with Michael that it also makes sense to have this logic outside the components, it needs to be executed only after the store is created, and it has nothing to do with the render cycle, it doesn't need anything but the URL, so it could be done outside. 
However, it is not clear to me where to encapsulate this logic, I don't think we have any action dispatch outside the components right now, If it doesn't exist perhaps we should start discussing an extensible architecture for it. 

Maybe we can leave this logic as a hook by now, and start conversations to create a framework for this kind of logic. What do you think @michaelolo24 ?